packages/redux-routine/src/index.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import isGenerator from './is-generator';
+
+/**
+ * Creates a Redux middleware, given an object of controls where each key is an
+ * action type for which to act upon, the value a function which returns either
+ * a promise which is to resolve when evaluation of the action should continue,
+ * or a value. The value or resolved promise value is assigned on the return
+ * value of the yield assignment. If the control handler returns undefined, the
+ * execution is not continued.
+ *
+ * @param {Object} controls Object of control handlers.
+ *
+ * @return {Function} Redux middleware function.
+ */
+export default function createMiddleware( controls = {} ) {
+	return ( store ) => ( next ) => ( action ) => {
+		if ( ! isGenerator( action ) ) {
+			return next( action );
+		}
+
+		function step( nextAction ) {
+			if ( ! nextAction ) {
+				return;
+			}
+
+			const control = controls[ nextAction.type ];
+			if ( typeof control === 'function' ) {
+				const routine = control( nextAction );
+
+				if ( routine instanceof Promise ) {
+					// Async control routine awaits resolution.
+					routine.then( ( result ) => {
+						step( action.next( result ).value );
+					} );
+				} else if ( routine !== undefined ) {
+					// Sync control routine steps synchronously.
+					step( action.next( routine ).value );
+				}
+			} else {
+				// Uncontrolled action is dispatched.
+				store.dispatch( nextAction );
+				step( action.next().value );
+			}
+		}
+
+		step( action.next().value );
+	};
+}
```

Comment:
I'd personally be in favor of using `rungen` just because it already handles all what we need:

 - recursion
 - raising errors when necessary
 - I'm fine with simplifying the `controls` API and making it "action type based" like in this middleware, this should be easy to do with `rungen` you just wrap the given controls in the middleware to rewrite them in the expected API.